File: lecode84.cpp
#include <vector>
#include <stack>
using namespace std;
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = heights.size();
        vector<int> left(n,-1),right(n,n);
        {stack<int> stk;
        for(int i = 0;i < heights.size();i++){
            if(stk.size() && heights[stk.top()] > heights[i]){
                while(stk.size() && heights[stk.top()] > heights[i]){
                    right[stk.top()] = i;
                    stk.pop();
                }
            }
            stk.push(i);
        }
        }

        {
        stack<int> stk;
        for(int i = n - 1;~i;i--){
             if(stk.size() && heights[stk.top()] > heights[i]){
                while(stk.size() && heights[stk.top()] > heights[i]){
                    left[stk.top()] = i;
                    stk.pop();
                }
            }
                stk.push(i);
        }
        }
        int res = 0;
        for(int i = 0;i < heights.size();i++){
            res = std::max(res,heights[i] * (right[i] - left[i] - 1));
        }
        return res;
    }
};  

File: lecode84.py
from typing import List
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        left = [-1] * n
        right = [n] * n
        stk = []
        for i in range(n):
            if stk and heights[stk[-1]] > heights[i]:
                while(stk and heights[stk[-1]] > heights[i]):
                    right [stk[-1]] = i
                    stk.pop()
            stk.append(i)
        stk.clear()
        for i in range(n - 1,-1,-1):
            if stk and heights[stk[-1]] > heights[i]:
                while(stk and heights[stk[-1]] > heights[i]):
                    left[stk[-1]] = i
                    stk.pop()
            stk.append(i)
        res = 0
        for i in range(n):
            res = max(res,heights[i] * (right[i] - left[i] - 1))
        return res